llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -0,0 +1,242 @@
+      reassociationIndices.push_back({i, i + 1});
+    } else if (i > splitDimParallel) {
+      reassociationIndices.push_back({i + 1});
+    }
+  }
+  Value valuesExpanded = rewriter.create<tensor::ExpandShapeOp>(
+      loc, valuesExpandedType, valuesOrig, reassociationIndices);
+
+  // Define the expanded output types
+  SmallVector<int64_t> expandedResultShape = expandedShape;
+  expandedResultShape[kDimParallel] = kSize;
+  RankedTensorType outputValuesExpandedType =
+      RankedTensorType::get(expandedResultShape, valueElementType);
+  RankedTensorType outputIndicesExpandedType =
+      RankedTensorType::get(expandedResultShape, indicesElementType);
+  SmallVector<Value> dynSizes;
+  for (auto en : llvm::enumerate(valuesExpandedType.getShape())) {
+    if (en.value() == ShapedType::kDynamicSize) {
+      dynSizes.push_back(
+          rewriter.create<tensor::DimOp>(loc, valuesExpanded, en.index()));
+    }
+  }
+
+  // Initialize the expanded output values
+  Value initTensorOutputValues = rewriter.create<mlir::linalg::InitTensorOp>(
+      loc, dynSizes, outputValuesExpandedType.getShape(), valueElementType);
+  Value initTensorOutputIndices = rewriter.create<mlir::linalg::InitTensorOp>(
+      loc, dynSizes, outputIndicesExpandedType.getShape(), indicesElementType);
+  // Initialize indices to positive infinity and values to negative infinity
+  // for a top (maxk) comparison.
+  Attribute negInfAttr;
+  if (auto intType = valueElementType.dyn_cast<IntegerType>()) {
+    negInfAttr = rewriter.getIntegerAttr(
+        intType, APInt::getSignedMinValue(intType.getWidth()));
+  } else {
+    auto negApFloat =
+        APFloat::getInf(valueElementType.cast<FloatType>().getFloatSemantics(),
+                        /*Negative=*/true);
+    negInfAttr = rewriter.getFloatAttr(valueElementType, negApFloat);
+  }
+  Value negInf = rewriter.create<arith::ConstantOp>(loc, negInfAttr);
+  Attribute posInfAttr =
+      rewriter.getIntegerAttr(indicesElementType, APInt::getSignedMaxValue(32));
+  Value posInf = rewriter.create<arith::ConstantOp>(loc, posInfAttr);
+  Value negInfTensor =
+      rewriter.create<linalg::FillOp>(loc, negInf, initTensorOutputValues)
+          .result();
+  Value posInfTensor =
+      rewriter.create<linalg::FillOp>(loc, posInf, initTensorOutputIndices)
+          .result();
+
+  // Parallel topk (first phase)
+  auto parallelTopkOp = rewriter.create<IREE::LinalgExt::TopkOp>(
+      loc,
+      /* resultTypes= */
+      TypeRange{outputValuesExpandedType, outputIndicesExpandedType},
+      /* ins= */ ValueRange{valuesExpanded},
+      /* outs= */ ValueRange{negInfTensor, posInfTensor}, kDimParallel);
+  rewriter.cloneRegionBefore(topkOp.region(), parallelTopkOp.region(),
+                             parallelTopkOp.region().end());
+
+  // Update the output indices from the parallel TopK with the reduction ratio
+  // offsets.
+  Value parallelIndices = parallelTopkOp.getResult(1);
```

Comment:
What is this part doing? Why is there a `linalg.generic` being created here?

Reply:
updated the comment to be fully descriptive.